File: src/main/kotlin/com/displee/compress/LZMACompressor.java
package com.displee.compress;

import com.displee.io.impl.InputBuffer;
import com.displee.io.impl.OutputBuffer;
import com.displee.util.Utils;
import lzma.sdk.lzma.Decoder;
import lzma.sdk.lzma.Encoder;
import lzma.streams.LzmaEncoderWrapper;

import java.io.*;


/**
 * A class handling LZMA compression.
 * @author Displee (credits to Techdaan)
 */
public class LZMACompressor {

	private static final Decoder DECODER = new Decoder();
	private static final Encoder ENCODER = new Encoder();
	private static final LzmaEncoderWrapper ENCODER_WRAPPER = new LzmaEncoderWrapper(ENCODER) {
		@Override
		public void code(InputStream in, OutputStream out) throws IOException {
			ENCODER.writeCoderProperties(out);
			ENCODER.code(in, out, -1, -1, null);
		}
	};

	static {
		ENCODER.setDictionarySize(8 << 20);
		ENCODER.setNumFastBytes(110);
		ENCODER.setMatchFinder(0);
		ENCODER.setLcLpPb(3, 0, 2);
		ENCODER.setEndMarkerMode(false);
	}

	public static byte[] compress(byte[] decompressed) {
		try {
			ByteArrayOutputStream baos = new ByteArrayOutputStream();
			ByteArrayInputStream bais = new ByteArrayInputStream(decompressed);
			lzma.streams.LzmaOutputStream lzma = new lzma.streams.LzmaOutputStream(baos, ENCODER_WRAPPER);
			Utils.INSTANCE.copy(bais, lzma);
			baos.close();
			lzma.close();
			bais.close();
			return baos.toByteArray();
		} catch (IOException e) {
			e.printStackTrace();
			return null;
		}
	}

	public static byte[] decompress(InputBuffer buffer, int decompressedLength) {
		OutputBuffer output = new OutputBuffer(buffer.remaining());
		output.write(buffer.raw(), buffer.getOffset(), buffer.remaining());
		return decompress(output.raw(), decompressedLength);
	}

	public static byte[] decompress(byte[] compressed, int decompressedLength) {
		try {
			return decompress(new ByteArrayInputStream(compressed), decompressedLength);
		} catch (IOException e) {
			e.printStackTrace();
			return null;
		}
	}

	public static byte[] decompress(ByteArrayInputStream input, int decompressedLength) throws IOException {
		byte[] properties = new byte[5];
		if (input.read(properties, 0, 5) != 5) {
			throw new IOException("LZMA: Bad input.");
		}
		ByteArrayOutputStream output = new ByteArrayOutputStream(decompressedLength);
		synchronized (DECODER) {
			if (!DECODER.setDecoderProperties(properties)) {
				throw new IOException("LZMA: Bad properties.");
			}
			DECODER.code(input, output, decompressedLength);
		}
		output.flush();
		return output.toByteArray();
	}

}

File: src/main/kotlin/com/displee/compress/type/LZMACompressor.kt
package com.displee.compress.type

import com.displee.io.impl.InputBuffer
import com.displee.io.impl.OutputBuffer
import com.displee.util.writeTo
import lzma.sdk.lzma.Decoder
import lzma.sdk.lzma.Encoder
import lzma.streams.LzmaEncoderWrapper
import lzma.streams.LzmaOutputStream
import java.io.*

/**
 * A class handling LZMA compression.
 * @author Displee (credits to Techdaan)
 */
class LZMACompressor : Compressor {

    private val DECODER = Decoder()
    private val ENCODER = Encoder()
    private val ENCODER_WRAPPER: LzmaEncoderWrapper = object : LzmaEncoderWrapper(ENCODER) {
        @Throws(IOException::class)
        override fun code(inputStream: InputStream, outputStream: OutputStream) {
            ENCODER.writeCoderProperties(outputStream)
            ENCODER.code(inputStream, outputStream, -1, -1, null)
        }
    }

    init {
        ENCODER.setDictionarySize(8 shl 20)
        ENCODER.setNumFastBytes(110)
        ENCODER.setMatchFinder(0)
        ENCODER.setLcLpPb(3, 0, 2)
        ENCODER.setEndMarkerMode(false)
    }

    override fun compress(bytes: ByteArray): ByteArray {
        val baos = ByteArrayOutputStream()
        try {
            val bais = ByteArrayInputStream(bytes)
            val lzma = LzmaOutputStream(baos, ENCODER_WRAPPER)
            bais.writeTo(lzma)
            baos.close()
            lzma.close()
            bais.close()
        } catch (e: IOException) {
            e.printStackTrace()
        }
        return baos.toByteArray()
    }

    override fun decompress(buffer: InputBuffer, compressedSize: Int, decompressedSize: Int): ByteArray {
        val output = OutputBuffer(buffer.remaining())
        output.writeBytes(buffer.raw(), buffer.offset, buffer.remaining())
        return try {
            decompress(ByteArrayInputStream(output.raw()), decompressedSize)
        } catch (e: IOException) {
            e.printStackTrace()
            byteArrayOf()
        }
    }

    @Throws(IOException::class)
    fun decompress(input: ByteArrayInputStream, decompressedLength: Int): ByteArray {
        val properties = ByteArray(5)
        if (input.read(properties) != 5) {
            throw IOException("LZMA: Bad input.")
        }
        val output = ByteArrayOutputStream(decompressedLength)
        synchronized(DECODER) {
            if (!DECODER.setDecoderProperties(properties)) {
                throw IOException("LZMA: Bad properties.")
            }
            DECODER.code(input, output, decompressedLength.toLong())
        }
        output.flush()
        return output.toByteArray()
    }

}